plugins/out_s3/s3.c

```diff
@@ -101,6 +115,44 @@ struct flb_aws_header *create_canned_acl_header(char *canned_acl)
     return acl_header;
 };
 
+static struct flb_aws_header *create_headers(struct flb_s3 *ctx, int num_headers)
+{
+    int n = 0;
+    struct flb_aws_header *canned_acl_header = NULL;
+    struct flb_aws_header *headers = NULL;
+    if (num_headers == 0) {
+        return NULL;
+    }
+
+    headers = flb_malloc(sizeof(struct flb_aws_header) * num_headers);
+    if (headers == NULL) {
+        flb_errno();
+        return NULL;
+    }
+
+    if (ctx->content_type != NULL) {
+        memcpy(&headers[n], &content_type_header, sizeof(content_type_header));
+        headers[n].val = ctx->content_type;
+        headers[n].val_len = strlen(ctx->content_type);
+        n++;
+    }
+    if (ctx->compression != NULL) {
+        headers[n] = content_encoding_header;
+        n++;
+    }
+    if (ctx->canned_acl != NULL) {
+        canned_acl_header = create_canned_acl_header(ctx->canned_acl);
+        if (canned_acl_header == NULL) {
+            flb_plg_error(ctx->ins, "Failed to create canned ACL header");
+            flb_free(headers);
+            return NULL;
+        }
+        headers[n] = *canned_acl_header;
```

Comment:
See comment, this should use memcpy too, and should probably be refactored to work the same way as the content_type header.